### ISSUE TITLE
A cached dependency doesn't call its constructor

### ISSUE THREAD

**Issue by akkie** [ray-di/Ray.Di#34]
I have an implementation which checks in its constructor if a static property is set or not. Based on this check the property gets initialized with an object instance. It's an simple singleton check. Now after the instance gets retrieved from cache the property is always null, because the constructor will never been executed.

``` php
class EventBusImpl implements EventBus {

    private static $dispatcher = null;

    public function __construct() {
        if (self::$dispatcher === null) {
            self::$dispatcher = new DefaultEventDispatcher();
        }
    }

    public function getDispatcher() {
        return self::$dispatcher;
    }
}
```

The method getDispatcher returns always null.

The binding its the following:

``` php
$this->bind('\util\EventBus')->to('\util\EventBusImpl');
```

**Comment by koriym** [comment]
Yes, cached object no longer uses the constructor or destructor. Instead, `sleep` and `wakeup` or Serializable interface is used.

So in this case you can either use `wakeup` or `unserialize`.

You probably already know that using static in the way you have described is a rare use case.

So therefore you can inject the dependency as below rather than instantiate in the constructor.

``` php
$this->bind('EventDispatcherInterface')->to('DefaultEventDispatcher')->int(Scope::SINGLETON);
```

**Comment by koriym** [comment]
@akkie There is another way using `@PostConstruct`.

`@PostConstruct` method called after all `@Inject`.

I should introduce this first :-/

``` php
use Ray\Di\Di\PostConstruct;

class EventBusImpl implements EventBus {

    private static $dispatcher = null;

    /**
     * @PostConstruct
     */
    public function init()
    {
        if (self::$dispatcher === null) {
            self::$dispatcher = new DefaultEventDispatcher();
        }
    }
```

**Comment by akkie** [comment]
Hey koriym,

I have now disabled caching of objects, because it leads to unexpected side effects. In another case I cache results of a DB query in a local property. If I enable caching then this local state is at once a shared global state.

Here is a short snippet to demonstrate the behavior:

``` php
class UserRepository {

    private $userId = null;

    private $cache = array();

    public function __construct($userId) {
        $this->userId = $userId;
    }

    public function getUserData() {
        if ($this->cache['userData']) {
            return $this->cache['userData'];
        }

        // Query
    }

    // Other user data
}

```

**Comment by koriym** [comment]
Hi @akkie , By the nature of object caching, You can't pass the data which changes per request  to constructor, it is reused beyond request.

**Comment by koriym** [comment]
This is example of cache-able object which I usually use in my system BEAR.Sunday.

``` php
class UserRepository {

    private $dependency1;
    private $dependency2;

    public function __construct()
    {
        // configure per system startup
    }

    /**
     * @PostConstruct
     */
    public function init()
    {
        // initialization in cache or non cache
    }

    public function getUserData($Id) {
        // request is stateless.
        // consumer (doesn't / can't) assume the object state.
    }
}

```

**Comment by akkie** [comment]
Yes I know. My problem was that I've upgraded from an old version of the lib(with the old caching behavior - not the object cache) to the newest version which introduces the new object cache. And I thought I can only change the initialization of the DI container with the new caching behavior. But then I experienced this addressed issues. Your implementation is OK and would I start from scratch I would implement it as you suggest it, but it's an old application and I will not spend to much time to fix such things. Therefore I have closed the issue.

Thanks for your help.
